pkg/scheduler/factory.go

```diff
@@ -210,143 +275,68 @@ func (c *Configurator) CreateFromConfig(policy schedulerapi.Policy) (*Scheduler,
 	if policy.AlwaysCheckAllPredicates {
 		c.alwaysCheckAllPredicates = policy.AlwaysCheckAllPredicates
 	}
-	return c.CreateFromKeys(predicateKeys, priorityKeys, extenders)
-}
 
-// CreateFromKeys creates a scheduler from a set of registered fit predicate keys and priority keys.
-func (c *Configurator) CreateFromKeys(predicateKeys, priorityKeys sets.String, extenders []algorithm.SchedulerExtender) (*Scheduler, error) {
 	klog.V(2).Infof("Creating scheduler with fit predicates '%v' and priority functions '%v'", predicateKeys, priorityKeys)
 
 	if c.hardPodAffinitySymmetricWeight < 1 || c.hardPodAffinitySymmetricWeight > 100 {
 		return nil, fmt.Errorf("invalid hardPodAffinitySymmetricWeight: %d, must be in the range 1-100", c.hardPodAffinitySymmetricWeight)
 	}
 
-	c.configProducerArgs.InterPodAffinityArgs = &interpodaffinity.Args{
+	args.InterPodAffinityArgs = &interpodaffinity.Args{
 		HardPodAffinityWeight: c.hardPodAffinitySymmetricWeight,
 	}
 
-	pluginsForPredicates, pluginConfigForPredicates, err := c.getPredicateConfigs(predicateKeys)
-	if err != nil {
-		return nil, err
-	}
-
-	pluginsForPriorities, pluginConfigForPriorities, err := c.getPriorityConfigs(priorityKeys)
+	pluginsForPredicates, pluginConfigForPredicates, err := getPredicateConfigs(predicateKeys, lr, args)
 	if err != nil {
 		return nil, err
 	}
 
-	priorityMetaProducer, err := getPriorityMetadataProducer(c.algorithmFactoryArgs)
+	pluginsForPriorities, pluginConfigForPriorities, err := getPriorityConfigs(priorityKeys, lr, args)
 	if err != nil {
 		return nil, err
 	}
-
 	// Combine all framework configurations. If this results in any duplication, framework
 	// instantiation should fail.
 	var plugins schedulerapi.Plugins
 	plugins.Append(pluginsForPredicates)
 	plugins.Append(pluginsForPriorities)
 	plugins.Append(c.plugins)
+	c.plugins = &plugins
```

Comment:
It seems `c.plugins` have to be appended again to itself, how about:

```go
if c.plugins == nil {
	c.plugins = &schedulerapi.Plugins{}
}
c.plugins.Append(pluginsForPredicates)
c.plugins.Append(pluginsForPriorities)
```

Reply:
This doesn't work because order matter, we need it to be last.